### PR TITLE
[Issue#76] Add MySQL test cases for threadpool

### PR DESCRIPTION
**Pull request by zwang28**
Add MySQL test cases for threadpool,
- threadpool_too_busy_active
- threadpool_too_busy_active-debug
- threadpool_too_busy_wait
- threadpool_too_busy_wait-debug
- threadpool_new_connection
- threadpool_pool_size
- threadpool_too_busy_corner_case

### CODE REVIEW

File: sql/threadpool_unix.cc [hunk 1]
```diff
@@ -668,6 +668,24 @@ static connection_t * listener(worker_thread_t *current_thread,
   
     cnt = io_poll_wait(thread_group->pollfd, ev, MAX_EVENTS, -1);
     TP_INCREMENT_GROUP_COUNTER(thread_group, polls[LISTENER]);
+    DBUG_EXECUTE_IF("threadpool_io_poll_wait_until_queue_not_empty",
+      mysql_mutex_lock(&thread_group->mutex);
```
Comment (zhang-lujie) [inline sql/threadpool_unix.cc:672]: DBUG_EXECUTE_IF的第二个参数是否整体包一个大括号比较方便阅读
DBUG_EXECUTE_IF("xxx",
  {
    xxx;
  };);

File: sql/threadpool_unix.cc [hunk 2]
```diff
@@ -668,6 +668,24 @@ static connection_t * listener(worker_thread_t *current_thread,
...
+      while (thread_group->queue.is_empty())
+      {
+        mysql_mutex_unlock(&thread_group->mutex);
+        my_sleep(1000000);
+        mysql_mutex_lock(&thread_group->mutex);
+      }
+      mysql_mutex_unlock(&thread_group->mutex);
+    );
+    DBUG_EXECUTE_IF("threadpool_io_poll_wait_at_lease_2_events",
```
Comment (zhang-lujie) [inline sql/threadpool_unix.cc:681]: at least

**Review by rejohn — APPROVED** [review]

File: sql/threadpool_unix.cc (hunk 1 above)
Comment (zwang28) [reply sql/threadpool_unix.cc]: 已修改

File: sql/threadpool_unix.cc (hunk 2 above)
Comment (zwang28) [reply sql/threadpool_unix.cc]: 已修改

File: mysql-test/t/threadpool_too_busy_wait.test [hunk 3]
```diff
@@ -0,0 +1,63 @@
+# Start with thread_handling=pool-of-threads
+
+--source include/have_pool_of_threads.inc
+
+# initialzie data
```
Comment (zhang-lujie) [inline mysql-test/t/threadpool_too_busy_wait.test:5]: initialize

Comment (zwang28) [reply mysql-test/t/threadpool_too_busy_wait.test]: 已修改

**Review by liuwei-ck — APPROVED** [review]

File: sql/threadpool_unix.cc [hunk 4]
```diff
@@ -668,6 +668,24 @@ static connection_t * listener(worker_thread_t *current_thread,
   
     cnt = io_poll_wait(thread_group->pollfd, ev, MAX_EVENTS, -1);
     TP_INCREMENT_GROUP_COUNTER(thread_group, polls[LISTENER]);
+    DBUG_EXECUTE_IF("threadpool_io_poll_wait_until_queue_not_empty",
+      mysql_mutex_lock(&thread_group->mutex);
+      while (thread_group->queue.is_empty())
```
Comment (yibantianxia) [inline sql/threadpool_unix.cc]: while后面的{不要新起一行

File: sql/threadpool_unix.cc [hunk 5]
```diff
@@ -668,6 +668,24 @@ static connection_t * listener(worker_thread_t *current_thread,
...
+      {
+        mysql_mutex_unlock(&thread_group->mutex);
+        my_sleep(1000000);
```
Comment (yibantianxia) [inline sql/threadpool_unix.cc]: 1000000算是神奇数

File: sql/threadpool_unix.cc [hunk 6]
```diff
@@ -668,6 +668,24 @@ static connection_t * listener(worker_thread_t *current_thread,
...
+    );
+    DBUG_EXECUTE_IF("threadpool_io_poll_wait_at_lease_2_events",
+      while (cnt < 2)
+      {
+        int cnt_again = io_poll_wait(thread_group->pollfd, ev + cnt, MAX_EVENTS - cnt, -1);
+        cnt += cnt_again;
+      }
+    );
+
     if (cnt <=0)
```
Comment (yibantianxia) [inline sql/threadpool_unix.cc:693]: <=后面要留空格

File: sql/threadpool_unix.cc [hunk 7]
```diff
@@ -669,20 +669,24 @@ static connection_t * listener(worker_thread_t *current_thread,
     cnt = io_poll_wait(thread_group->pollfd, ev, MAX_EVENTS, -1);
     TP_INCREMENT_GROUP_COUNTER(thread_group, polls[LISTENER]);
     DBUG_EXECUTE_IF("threadpool_io_poll_wait_until_queue_not_empty",
-      mysql_mutex_lock(&thread_group->mutex);
-      while (thread_group->queue.is_empty())
       {
```
Comment (yibantianxia) [inline sql/threadpool_unix.cc:672]: while后面的{不用另起一行

File: sql/threadpool_unix.cc [hunk 8]
```diff
@@ -669,20 +669,24 @@ static connection_t * listener(worker_thread_t *current_thread,
...
-        mysql_mutex_unlock(&thread_group->mutex);
-        my_sleep(1000000);
         mysql_mutex_lock(&thread_group->mutex);
+        while (thread_group->queue.is_empty())
+        {
+          mysql_mutex_unlock(&thread_group->mutex);
+          my_sleep(1000000);
```
Comment (yibantianxia) [inline sql/threadpool_unix.cc:677]: 神奇数1000000

**Pull request merged**